File: KwizzadRTASampleObjC/ViewController.m
//
//  ViewController.m
//  KwizzadRTASampleObjC
//
//  Created by Fares Ben Hamouda on 07.01.20.
//  Copyright © 2020 Fares Ben Hamouda. All rights reserved.
//

#import "ViewController.h"
#import <KwizzadRTA/KwizzadRTA-Swift.h>
@interface ViewController () <KwizzadRTADelegate>

@property (weak, nonatomic) IBOutlet UITextView *debugTextView;
@property (weak, nonatomic) IBOutlet UIButton *btnShowAd;

@end

@implementation ViewController

KwizzadRTA* kwizzad;
NSString* placement = @"test";

- (void) viewDidLoad {
    [super viewDidLoad];
    [self setTitle:@"KwizzadRTA Sample ObjC"];
    
    kwizzad = [[KwizzadRTA alloc]init];
}

- (IBAction)loadAdClicked:(id)sender {
    [kwizzad loadWithPlacement:@"test" delegate:self];
    [_debugTextView insertText: [NSString stringWithFormat:@"\n start loading placement %@", placement]];
}

- (IBAction)showAdClicked:(id)sender {
    [kwizzad showAdFrom:self];
}

- (IBAction)showDebugScreen:(id)sender {
    [KwizzadRTA showDebugScreenFrom:self instances: [NSArray arrayWithObjects: kwizzad, nil]];
}

- (IBAction)showConsentScreen:(id)sender {
    [KwizzadRTA showConsentScreenFrom:self instances: [NSArray arrayWithObjects: kwizzad, nil] completion:^(BOOL consentGiven) {
        [self->_debugTextView insertText: [NSString stringWithFormat:@"\n consent status: %@", consentGiven ? @"true" : @"false"]];
    }];
}

// MARK: KwizzadRTA Delegate

- (void)onAdAvailableWithPlacementId:(NSString * _Nonnull)placementId {
    [_btnShowAd setEnabled:YES];
    [_debugTextView insertText: [NSString stringWithFormat:@"\n an ad is preloaded on placement %@", placementId]];
}

- (void)onAdCanceledWithPlacementId:(NSString * _Nonnull)placementId params:(NSDictionary<NSString *,id> * _Nullable)params {
    [_debugTextView insertText: [NSString stringWithFormat:@"\n ad is cancelled on placement %@", placementId]];
}

- (void)onAdErrorWithPlacementId:(NSString * _Nonnull)placementId error:(NSString * _Nonnull)error params:(NSDictionary<NSString *,id> * _Nullable)params {
    [_debugTextView insertText: [NSString stringWithFormat:@"\n an error occured on placement %@, error: %@", placementId, error]];
}

- (void)onAdFailedToLoadWithPlacementId:(NSString * _Nonnull)placementId {
    [_debugTextView insertText: [NSString stringWithFormat:@"\n ad failed to load on placement %@", placementId]];

}

- (void)onAdFinishedWithPlacementId:(NSString * _Nonnull)placementId params:(NSDictionary<NSString *,id> * _Nullable)params {
    [_debugTextView insertText: [NSString stringWithFormat:@"\n ad is finished on placement %@", placementId]];
}

- (void)onAdPassbackWithPlacementId:(NSString * _Nonnull)placementId params:(NSDictionary<NSString *,id> * _Nullable)params {
    [_debugTextView insertText: [NSString stringWithFormat:@"\n a passback occured on placement %@", placementId]];
}


@end

File: KwizzadRTASample/ViewController.swift
//
//  ViewController.swift
//  KwizzadRTASample
//
//  Created by Fares Ben Hamouda on 25.10.19.
//  Copyright © 2019 Fares Ben Hamouda. All rights reserved.
//

import UIKit
import KwizzadRTA

class ViewController: UIViewController {

    var kwizzad: KwizzadPlacement?

    @IBOutlet weak var buttonShowAd: UIButton!
    @IBOutlet weak var textViewDebug: UITextView!
    @IBOutlet weak var labelVersion: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.title = "KwizzadRTA Sample"
        labelVersion.text?.append(KwizzadRTA.sdkVersion())
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.kwizzad = KwizzadPlacement(with: placement, delegate: self)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        DispatchQueue.main.asyncAfter(deadline: .now()+1) {
            self.kwizzad?.load()
            self.textViewDebug.appendText(with: "start loading placement \(placement)")
        }
    }
    
    @IBAction func loadAdClicked(_ sender: Any) {
        self.kwizzad?.preloadAdsManually()
        self.textViewDebug.appendText(with: "reloading placement \(placement)")
    }
    
    @IBAction func showAdClicked(_ sender: Any) {
        KwizzadRTA.setCustomParams(["test": "test"])
        kwizzad?.showAd(from: self)
    }

    @IBAction func showDebugScreen(_ sender: Any) {
        guard let kwizzad = kwizzad else {return}
        KwizzadRTA.showDebugScreen(from: self, instances: [kwizzad])
    }
    
    @IBAction func showConsentScreen(_ sender: Any) {
        guard let kwizzad = kwizzad else {return}
        KwizzadRTA.showConsentScreen(from: self, instances: [kwizzad]) { (consentGiven) in
            self.textViewDebug.appendText(with: "consent status: \(consentGiven)")
        }
    }
}

extension ViewController: KwizzadRTADelegate {
        
    func onAdAvailable(placementId: String) {
        buttonShowAd.isEnabled = true
        self.textViewDebug.appendText(with: "an ad is preloaded on placement \(placementId)")
    }
    
    func onAdFailedToLoad(placementId: String) {
        self.textViewDebug.appendText(with: "ad failed to load on placement \(placementId)")
    }

    func onAdCanceled(placementId: String, params: [String : Any]?) {
        self.textViewDebug.appendText(with: "ad is cancelled on placement \(placementId)")
    }
    
    func onAdError(placementId: String, error: String, params: [String : Any]?) {
        self.textViewDebug.appendText(with: "an error occured on placement \(placementId), error: \(error)")
    }
    
    func onAdPassback(placementId: String, params: [String : Any]?) {
        self.textViewDebug.appendText(with: "a passback occured on placement \(placementId)")
    }
    
    func onAdFinished(placementId: String, params: [String : Any]?) {
        self.textViewDebug.appendText(with: "ad is finished on placement \(placementId)")
    }
    
    //optionals callbacks
    func onAdOpened(placementId: String) {}
    func onAdClicked(placementId: String) {}
    func onConsentShouldBeUpdated(placementId: String) {}
}

// MARK: Helpers

extension UITextView {
    
    func appendText(with text: String) {
        self.text.append("\n"+text)
    }
    
}

